src/middle/Middle.ml

```diff
@@ -143,8 +145,12 @@ let rec infer_type_of_indexed ut indices =
    |UMatrix, [Upfrom _; Single _]
    |UMatrix, [Between _; Single _]
    |UMatrix, [MultiIndex _]
-   |UMatrix, [Single _] ->
-      UVector
+   |UMatrix, [Single _] -> UVector
+  | USparseMatrix, [All; Single _]
+   |USparseMatrix, [Upfrom _; Single _]
+   |USparseMatrix, [Between _; Single _]
+   |USparseMatrix, [MultiIndex _; MultiIndex _; All; All;]
+   |USparseMatrix, [Single _] -> UVector
```

Comment:
Can you explain the `index` types like `Single`? I can't really understand what they do or relate to?

Reply:
They correspond to these: https://mc-stan.org/docs/2_20/reference-manual/language-indexing-section.html

You can check out how it gets parsed in [the parser](https://mc-stan.org/docs/2_20/reference-manual/language-indexing-section.html)

Single - regular old C style index, a[1]
Upfrom - a[1:]
Between - a[1:5]
MultiIndex - indexing by an int array of indices, a[{1, 2, 5}]


That all is probably irrelevant as I'm pretty sure Dan Simpson wanted to disallow indexing of sparse matrices, right?

Reply:
In `mkfor` it assums the loop is happening over a `Single` and I wasn't sure exactly which of these that lined up with. Thanks!